### ISSUE TITLE
Overriding default currency rounding

### ISSUE THREAD

**Issue by morinap** [kipcole9/money#122]
I'm working with money values that include fractional cents and would like to easily display those values universally across my Phoenix application. I understand I can pass `fractional_digits: 4` to `Money.to_string!` calls to get the display value I'm seeking, but ideally I'd like to be able to either override that in the currency definition, or to define that value in the protocol definitions found in `https://github.com/kipcole9/money/blob/master/lib/money/protocol_implementations.ex` so that when I display a Money value in a rendered Phoenix page it "just works" (`Phoenix.HTML.Safe`).

I can redefine `defimpl String.Chars, for: Money` and `defimpl Phoenix.HTML.Safe, for: Money` in my application and update the `to_string` calls to pass my desired `fractional_digits` value, but that gives me module redefinition compiler warnings and also seems like the incorrect way to approach this.

Am I thinking about this incorrectly or is there a different way to do this? I'm also willing to fork and open a PR back to you to allow this value to be specified in the Cldr backend configuration, but I'm not sure that's the approach you'd like to see, either.

**Comment by kipcole9** [comment]
Thats for opening the issue, what you describes make sense. I also had a similar requirement in [ex_cldr_units](https://github.com/elixir-cldr/cldr_units) which I addressed by adding a `:format_options` to the struct and similarly allowing format options to be added in `Money.new!/2` and friends.

That way your would be calling `Money.new!(:USD, 100, fractional_digits: 4)` or something similar.

Would that work for you in principle?  The required changed would be:

* Add `:format_options` to to `t:Money`
* Update `Money.to_string/2` and friends to honour the options. Any format options to `Money.to_string/2` would be merged into any `:format_options` in the struct.

One possible issue is that serialising to the database would lose the format options since the composite type does not include them today, and I'm not sure that it should. May require a mechanism to add format options to a `Money`. Any thoughts?

I'll be able to complete this over the weekend if there are no major unanticipated hurdles.

**Comment by morinap** [comment]
Thank you for the quick reply. Your suggestion makes sense as a general case, but in my situation it may not be as helpful as the majority of the time I'm displaying these fractional values, they are `Money` structs serialized from a Postgres database. It would be a more extensive change, but your proposal could be coupled with altering [Money.Ecto.Composite.Type|https://github.com/kipcole9/money_sql/blob/master/lib/money/ecto/money_ecto_composite_type.ex] to be a [ParameterizedType|https://hexdocs.pm/ecto/Ecto.ParameterizedType.html] and allow the fractional digits to be specified as a field parameter and populated on the struct whenever `load` is called.

Again, I'm also willing to make changes myself and submit a pull request back to either or both projects if you're amenable to that, but if you'd prefer to make changes yourself or not to make changes at all I completely understand.

**Comment by kipcole9** [comment]
Thanks for the update. I'm not averse to updating the composite type (or providing a new one). My concern is the overall performance of serialising format options to json and back. Or I suppose it could even be serialised with `:erlang.term_to_binary/2`.  Is that going to be acceptable in your case?

**Comment by morinap** [comment]
Perhaps I'm misunderstanding your comment or how Parameterized Types work, but based on my reading of the docs, the field parameters are initialized at compile time and passed to the type callbacks, not stored in the database. I agree with you that storing this sort of configuration in the DB would be both inefficient and an improper separation of duties.

**Comment by kipcole9** [comment]
Ah, if parameterised type would be "enough" for you then that would be much simpler since it wouldn't require any update to the Postgres composite type since the format options would be applied in the Ecto type.

Two approaches forward then:
1. Change the composite type so that format options are serialised to the data base along with the currency and value. That means format options can be added per money instance.
2. Add a parameterised type in Ecto whereby the format options are applied through the schema. No changes to the underlying database, but an updated `Money.Ecto.Composite.Type` that works as a parameterised type.

Sounds like (2) is ok for you?

**Comment by kipcole9** [comment]
Have started a feature branch and updated the struct in [this commit](https://github.com/kipcole9/money/commit/ec49ee6b91c84ae81b96fd296cc17e3bb54e1253)

**Comment by kipcole9** [comment]
Proof of concept suggest this works well.  However I need to decide what to do with money operations (plus, minus, sum, ...) where a new money is returned. I see the primary options as:

1. Merge `:format_options` of each operand. Not preferred since it produces unpredictable results.
2. Use the `:format_options` of the second operand.  ie if we have `Money.add(money_1, money_2)` the resulting money would have the format options from `money_2`.

Thoughts?

**Comment by morinap** [comment]
Excellent, thank you for the quick work! I agree that #2 is the preferred choice, using the `:format_options` from the second operand. This is consistent with similar behavior from existing Elixir functions such as `Map.merge`.

**Comment by kipcole9** [comment]
As of [this commit](https://github.com/kipcole9/money/commit/be4e29cd5b7f8495f3a82b1b30fb3dbba5109b16), `:format_options` is propagate through all operations although the test of this is not 100% coverage yet.  If you're feeling adventurous, the master branch is ready for early exposure.  I'll work on the parameterised type next.

**Comment by kipcole9** [comment]
As of [this commit](https://github.com/kipcole9/money_sql/commit/5c6d00ee2a2d2845725d6ebfc7c6dc5d7e19b122) in [money_sql](https://github.com/kipcole9/money_sql), the `Money.Ecto.Composite.Type` is now ` ParamaterizedType`.  There are a couple of other issues there I need to resolve. Subject to any feedback, I will aim to publish in new releases in the next couple of days.

**Comment by kipcole9** [comment]
I've published [money version 5.5.0](https://hex.pm/packages/ex_money).  I will close the issue for now but please reopen if there are any undetected issues you find. Thanks for the collaboration and support!

**Comment by morinap** [comment]
@kipcole9 Thank you for such a quick turnaround, I'll pull it into my repo and give it a try!

**Comment by kipcole9** [comment]
Forgot to mention that I also published [ex_money_sql version 1.4.0](https://hex.pm/packages/ex_money_sql/1.4.0) that includes the change to a parameterized type and therefore can be configured in the schema to apply `:formatting_options` so I think - I hope - you are now good to go.

Of course reopen the issue if you see something wrong.

**Comment by morinap** [comment]
@kipcole9 I have found one minor thing that I think is best fixed in `money_sql`, although it may be a problem in `ecto` itself. Ecto defines multiple patterns for `cast`, including one that just returns `nil` if `nil` is cast (see [here](https://github.com/elixir-ecto/ecto/blob/3e796615c7a2ebd6961f576fda5d75faf85799a1/lib/ecto/type.ex#L739)). No such pattern exists for a parameterized cast of a `nil` value, and `money_sql` will return `:error` if a `nil` value is cast (see [here](https://github.com/kipcole9/money_sql/blob/324c6919203a70f14d70b34dd7ea5894b3f0a53a/lib/money/ecto/money_ecto_composite_type.ex#L100)). My initial thought is that Ecto should have a pattern to return `nil` for a `nil` parameterized type, but upon further thinking it may be intentional that it's left up to the type as the parameters themselves may cause different type implementations to respond to `nil` values differently.

In summary, I believe `money_sql` should implement a `cast` pattern that returns `nil` for `nil` values. If you'd prefer me to open an issue in that repo, I can.

**Comment by kipcole9** [comment]
Thanks for the feedback, good call. Have published [money_sql version 1.4.1](https://hex.pm/packages/money_sql/1.4.1) that casts `nil` to `{:ok, nil}` in the composite type.

**Comment by morinap** [comment]
@kipcole9 I believe I missed that the same change needs to be applied on the `dump` side to allow `nil` values to be written back to the database. Currently that fails for me:
```
     ** (Ecto.ChangeError) value `nil` for `<redacted>` in `update` does not match type {:parameterized, Money.Ecto.Composite.Type, []}
     code: <redacted>
     stacktrace:
       (ecto 3.5.7) lib/ecto/repo/schema.ex:889: Ecto.Repo.Schema.dump_field!/6
       (ecto 3.5.7) lib/ecto/repo/schema.ex:898: anonymous fn/6 in Ecto.Repo.Schema.dump_fields!/5
       (stdlib 3.14) maps.erl:233: :maps.fold_1/3
       (ecto 3.5.7) lib/ecto/repo/schema.ex:896: Ecto.Repo.Schema.dump_fields!/5
       (ecto 3.5.7) lib/ecto/repo/schema.ex:829: Ecto.Repo.Schema.dump_changes!/6
       (ecto 3.5.7) lib/ecto/repo/schema.ex:334: anonymous fn/15 in Ecto.Repo.Schema.do_update/4
       <redacted>
```

**Comment by kipcole9** [comment]
Fixed in [ex_money_sql 1.4.2](https://hex.pm/packages/ex_money_sql/1.4.2). Have added tests for dumping/loading and added actual DB operations in tests to check. I think end-to-end it should be fine now. Sorry for the poor testing of this, not the quality I aim for.

**Comment by morinap** [comment]
No problem at all, I understand and appreciate the quick turnaround on all of this.
